src/main/java/edu/bu/metcs673/trackr/user/TrackrUserServiceImpl.java

```diff
@@ -92,24 +92,25 @@ public String createUser(TrackrUserDTO userInput) {
     public TrackrUserDTO updateUser(TrackrUserDTO dto) {
         TrackrUser trackrUser = getCurrentUser();
 
-        if (!bCryptPasswordEncoder.matches(dto.getPassword(), trackrUser.getPassword()) && StringUtils.isNotBlank(dto.getPassword())) {
-            throw new TrackrInputValidationException(CommonConstants.WRONG_PASSWORD);
+        if (StringUtils.isNotBlank(dto.getNewPassword())) {
+            if (!bCryptPasswordEncoder.matches(dto.getPassword(), trackrUser.getPassword())) {
+                throw new TrackrInputValidationException(CommonConstants.WRONG_PASSWORD);
+            } else {
+                trackrUser.setPassword(bCryptPasswordEncoder.encode(dto.getNewPassword()));
+            }
         }
 
         trackrUser.setFirstName(dto.getFirstName());
         trackrUser.setLastName(dto.getLastName());
         trackrUser.setEmail(dto.getEmail());
 
-        if (StringUtils.isNotBlank(dto.getNewPassword())) {
-            trackrUser.setPassword(bCryptPasswordEncoder.encode(dto.getNewPassword()));
-        }
-
         trackrUser = userRepository.save(trackrUser);
 
         TrackrUserDTO updated = new TrackrUserDTO();
         updated.setFirstName(trackrUser.getFirstName());
         updated.setLastName(trackrUser.getLastName());
         updated.setEmail(trackrUser.getEmail());
+        updated.setUsername(trackrUser.getUsername());
```

Comment:
We don't need to set the user name since it's available in the token. On the other side, you could say that we want to display it in the UI then it's ok since it's not sensitive information. The main gotcha is that we never pass the username in a request and only fetch it in the token.